File: examen-backend/src/main/java/com/rei/examenbackend/service/CategoryService.java
package com.rei.examenbackend.service;

import com.rei.examenbackend.dto.category.CategoryRequest;
import com.rei.examenbackend.dto.category.CategoryResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface CategoryService {

    CategoryResponse create(CategoryRequest request);

    CategoryResponse update(Long id, CategoryRequest request);

    void delete(Long id);

    CategoryResponse getById(Long id);

    Page<CategoryResponse> getAll(Pageable pageable);
}

File: examen-frontend/src/app/categories/category-list/services/category.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { ApiConfigService } from '../../../core/services/api-config.service';

export interface Category {
  id?: number;
  name: string;
  description: string;
}

interface PageResponse<T> {
  content: T[];
}

@Injectable({
  providedIn: 'root'
})
export class CategoryService {
  private baseUrl: string;

  constructor(
    private http: HttpClient,
    private api: ApiConfigService
  ) {
    this.baseUrl = this.api.endpoint('categories');
  }

  getAll(): Observable<Category[]> {
    return this.http
      .get<PageResponse<Category>>(this.baseUrl, { params: { page: 0, size: 200 } })
      .pipe(map(res => res.content));
  }

  getById(id: number): Observable<Category> {
    return this.http.get<Category>(`${this.baseUrl}/${id}`);
  }

  create(c: Category): Observable<Category> {
    return this.http.post<Category>(this.baseUrl, c);
  }

  update(id: number, c: Category): Observable<Category> {
    return this.http.put<Category>(`${this.baseUrl}/${id}`, c);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }
}
